zeebe/dynamic-node-id-provider/src/main/java/io/camunda/zeebe/dynamic/nodeid/RepositoryNodeIdProvider.java

```diff
@@ -0,0 +1,207 @@
+        if (retryRound > 1) {
+          try {
+            currentDelay = backoff.supplyRetryDelay(currentDelay);
+            LOG.debug(
+                "Attempt to acquire the lease failed for all nodeIds, sleeping {} and retrying again",
+                currentDelay);
+            Thread.sleep(currentDelay);
+          } catch (final InterruptedException e) {
+            break;
+          }
+        }
+      }
+      final var nodeId = i++ % clusterSize;
+      final var storedLease = nodeIdRepository.getLease(nodeId);
+      currentLease = tryAcquire(storedLease);
+    }
+    if (currentLease != null) {
+      LOG.info(
+          "Acquired lease w/ nodeId={}.  {}", currentLease.lease().nodeInstance(), currentLease);
+    } else {
+      throw new IllegalStateException("Failed to acquire a lease");
+    }
+  }
+
+  private StoredLease.Initialized tryAcquire(final StoredLease lease) {
+    try {
+      switch (lease) {
+        case final Initialized initialized -> {
+          if (initialized.lease().isStillValid(clock.millis(), leaseDuration)) {
+            LOG.debug("Lease {} is held by another process, skipping it", initialized);
+            return null;
+          } else {
+            LOG.debug("Trying to acquire an expired lease {}", initialized);
+            final var newLease =
+                (initialized.metadata().task().equals(taskId))
+                    ? initialized.lease().renew(clock.millis(), leaseDuration)
+                    : new Lease(
+                        taskId,
+                        clock.millis() + leaseDuration.toMillis(),
+                        initialized.lease().nodeInstance());
+            return nodeIdRepository.acquire(newLease, initialized.eTag());
+          }
+        }
+        case final Uninitialized uninitialized -> {
+          final var newLease =
+              new Lease(taskId, clock.millis() + leaseDuration.toMillis(), uninitialized.node());
+          LOG.debug(
+              "Trying to take uninitialized lease: {} with new lease {}", uninitialized, newLease);
+          return nodeIdRepository.acquire(newLease, uninitialized.eTag());
+        }
+      }
+    } catch (final Exception e) {
+      LOG.warn("Failed to acquire the lease {}", lease, e);
+      return null;
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    // use the executor status to avoid closing multiple times
+    if (!executor.isShutdown()) {
+      try {
+        if (renewalTask != null) {
+          renewalTask.cancel(true);
+        }
+        if (currentLease != null) {
+          nodeIdRepository.release(currentLease);
+        }
+      } finally {
+        currentLease = null;
+        // release is already submitted to the executor, we can shut it down gracefully.
+        executor.shutdown();
+        // shutdown is taking too much time, let's shut it down by interrupting running tasks.
+        if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
```

Comment:
❓ Could this make shutdown slow in some tests?

Reply:
In 99.9% of the tests, this NodeIdProvider is not used. The time taken to shutdown the executor is if a running task is blocking. 
We use this executor basically just to make S3 requests to `LocalStack` so it should be ~ 100-200 millis in most cases.

Reply:
Mostly because we are spawning just 1 thread and with jdk21 [JEP-491](https://openjdk.org/jeps/491) is not fixed yet. If you say we can ignore it, I can promote it to a virtual thread executor.